benchmark/conversions/conversions.cpp

```diff
@@ -0,0 +1,270 @@
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <algorithm>
+#include <chrono>
+#include <cstdlib>
+#include <exception>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <typeinfo>
+
+
+#include "benchmark/utils/general.hpp"
+#include "benchmark/utils/loggers.hpp"
+
+
+using etype = double;
+using duration_type = std::chrono::nanoseconds;
+using hybrid = gko::matrix::Hybrid<>;
+using csr = gko::matrix::Csr<>;
+
+
+// input validation
+void print_config_error_and_exit()
+{
+    std::cerr << "Input has to be a JSON array of matrix configurations:"
+              << "[\n    { \"filename\": \"my_file.mtx\"},"
+              << "\n    { \"filename\": \"my_file2.mtx\"}"
+              << "\n]" << std::endl;
+    std::exit(1);
+}
+
+
+void validate_option_object(const rapidjson::Value &value)
+{
+    if (!value.IsObject() || !value.HasMember("filename") ||
+        !value["filename"].IsString()) {
+        print_config_error_and_exit();
+    }
+}
+
+
+// Command-line arguments
+DEFINE_string(
+    formats, "coo",
+    "A comma-separated list of formats to run."
+    "Supported values are: coo, csr, ell, hybrid, sellp"
+    "coo: Coordinate storage.\n"
+    "csr: Compressed Sparse Row storage.\n"
+    "ell: Ellpack format according to Bell and Garland: Efficient Sparse "
+    "Matrix-Vector Multiplication on CUDA.\n"
+    "hybrid: Hybrid uses ell and coo to represent the matrix.\n"
+    "sellp: Sliced Ellpack format.\n");
+
```

Comment:
I understand correctly, the conversion is always COO -> *other format* -> COO ?

Reply:
When it is said that the user wants to benchmark conversions for a format (say csr), conversions are automatically tried against all other formats in Ginkgo (i.e. coo, ell, hybrid, sellp).

In other words, the users only give "format1" as argument to the benchmarker. Then, all possible format2 are tested. If both format1 and format2 are given as argument then both ways will be benchmarked. I thought if the users specify both format1 and format2 as arguments it is too tiresome as there is a lot of possible combinations.

Reply:
Makes sense. But we should be explicit in the description, then...